File: workspaces/ros2_ws/src/my_py_pkg/my_py_pkg/num_pub.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from example_interfaces.msg import Int64

class NumPubNode(Node):
    def __init__(self):
        super().__init__("number_publisher")

        # Set up parameters
        self.declare_parameter("number", 2)
        self.declare_parameter("timer_period", 1.0)

        # Set to attributes
        self.number_ = self.get_parameter("number").value
        self.timer_period_ = self.get_parameter("timer_period").value

        # Configure callback for runtime parameter update
        self.add_post_set_parameters_callback(self.parameters_callback)

        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(self.timer_period_, self.publish_number)
        self.get_logger().info("Num Pub has been started.")

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.publisher_.publish(msg)

    def parameters_callback(self, params: list[Parameter]):
        for param in params:
            if param.name == "number":
                self.number_ = param.value
            if param.name == "timer_period":
                self.timer_period_ = param.value
                self.timer_.cancel()
                self.timer_ = self.create_timer(self.timer_period_, self.publish_number)

def main(args=None):
    rclpy.init(args=args)
    node = NumPubNode()
    rclpy.spin(node)
    rclpy.shutdown()

File: workspaces/ros2_ws/src/my_cpp_pkg/src/num_pub.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node
{
public:
NumberPublisherNode() : Node("number_publisher")
    {

        // Get parameters
        this->declare_parameter("number", 2);
        this->declare_parameter("timer_period", 1.0);

        // Set parameters to member variables
        number_ = this->get_parameter("number").as_int();
        timer_period_ = this->get_parameter("timer_period").as_double();

        // Set parameter callback
        param_cb_handle_ = this->add_post_set_parameters_callback(
            std::bind(&NumberPublisherNode::parametersCallback, this, std::placeholders::_1)
        );

        // Create publisher
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(timer_period_),
            std::bind(&NumberPublisherNode::publishNumber, this)
        );
        RCLCPP_INFO(this->get_logger(), "Num Pub Node has been started.");
    }

private:
    void publishNumber()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        publisher_->publish(msg);
    }

    void parametersCallback(const std::vector<rclcpp::Parameter> &parameters)
    {
        for (const auto &param: parameters) {
            if (param.get_name() == "number") {
                number_ = param.as_int();
            }
            if (param.get_name() == "timer_period") {
                timer_period_ = param.as_double();
                timer_->cancel();
                timer_ = this->create_wall_timer(
                    std::chrono::duration<double>(timer_period_),
                    std::bind(&NumberPublisherNode::publishNumber, this)
                );
            }
        }
    }

    int number_;
    double timer_period_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::node_interfaces::PostSetParametersCallbackHandle::SharedPtr param_cb_handle_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
